api/apps/geoprocessing/src/legacy-project-import/legacy-piece-importers/file-readers/dat-file.comma-or-tab-finder.ts

```diff
@@ -0,0 +1,42 @@
+import { Injectable } from '@nestjs/common';
+import { Either, left, right } from 'fp-ts/lib/Either';
+import { createInterface } from 'readline';
+import { Readable } from 'typeorm/platform/PlatformTools';
+import {
+  DatFileDelimiterFinder,
+  invalidDelimiter,
+} from './dat-file.delimiter-finder';
+
+export type DatFileCommaOrTab = Either<typeof invalidDelimiter, ',' | '\t'>;
+
+@Injectable()
+export class DatFileCommaOrTabFinder implements DatFileDelimiterFinder {
+  async findDelimiter(file: Readable): Promise<DatFileCommaOrTab> {
+    const rl = createInterface({
+      input: file,
+    });
+
+    const result = await new Promise<DatFileCommaOrTab>((resolve) => {
+      rl.on(`line`, (line) => {
+        const { tabDelimiter, commaDelimiter } = this.commaOrTabDelimiter(line);
+
+        if (!tabDelimiter && !commaDelimiter) resolve(left(invalidDelimiter));
+
+        resolve(right(tabDelimiter ? '\t' : ','));
+      });
+    });
+
+    rl.close();
+    return result;
+  }
+
+  private commaOrTabDelimiter = (line: string) => {
+    const tabPattern = /[a-z]\t[a-z]+/;
+    const commaPattern = /[a-z],[a-z]+/;
+
+    const tabDelimiter = tabPattern.test(line);
+    const commaDelimiter = commaPattern.test(line);
+
+    return { tabDelimiter, commaDelimiter };
+  };
```

Comment:
This should work in the specific case of Marxan input files where users didn't do much messing-up of their files - in theory trying to parse (via `fast-csv`) the first line with both comma and tab as separators and check if we have "columns" once parsing should help with (for example) spurious tabs that may have sneaked into the header line of a comma-separated file, but yeah, let's keep this simple (especially as parsing will likely break later on in any case if we "guess" the separator wrong here).
Likewise if a file contains only one column we wouldn't recognize the separator here (but likely not by attempting to parse via `fast-csv` either, unless we handle the case explicitly), but again in practice there should never be any Marxan legacy input files with one single column, so tl;dr I think we're ok here - I wrote this comment up for you @angelhigueraacid  to cross-check, but I think we can merge as is.

Reply:
Yes, now thinking of it, perhaps we could have use the `fast-csv` library for reading the firs line and skipping the rest. We can always just make a new implementation for `DatFileDelimiterFinder` using `fast-csv`.
For dat files that only contain one column, the result will be `invalidDelimiter` so i think it should be okey as any input file with just one column seems already invalid. I will solve the `fisrtLine` typos that i spread around every legacy importer and merge the PR!